p2p/protocol.go

```diff
@@ -44,17 +44,31 @@ func NewProtocol(network Service, name string, requestTimeout time.Duration) *Pr
 		pendingMap:         make(map[uint64]chan interface{}),
 		resHandlers:        make(map[uint64]func(msg []byte)),
 		msgRequestHandlers: make(map[MessageType]func(msg []byte) []byte),
-		Exit:               make(chan struct{}),
+		exit:               make(chan struct{}),
 	}
 
 	go p.readLoop()
 	return p
 }
 
+func (p *Protocol) Close() {
+	p.exit <- struct{}{}
```

Comment:
I don't understand how it works, you can read from the channel once. so all the other loops waiting on `<-p.exit` will just block forever.
you should use `close(p.exit)` this lets you block on the channel and release in multiple places when closed.

Reply:
it had another issue of waiting for all gorutines to stop, but the idea is i want to wait for the read loop to exit before i close all the resources
add a waitgroup.wait to handle running gorutines